src/psyclone/psyir/tools/definition_use_chains.py

```diff
@@ -0,0 +1,629 @@
+                           dependency search.
+
+    :raises TypeError: If one of the arguments is the wrong type.
+
+    """
+
+    def __init__(
+        self,
+        reference,
+        control_flow_region=None,
+        is_local=None,
+        start_point=None,
+        stop_point=None,
+    ):
+        if not isinstance(reference, Reference):
+            raise TypeError(f"The reference passed into a DefinitionUseChain "
+                            f"must be a Reference but found "
+                            f"'{type(reference).__name__}'.")
+        self._reference = reference
+        # Store the absolute position for later.
+        self._reference_abs_pos = reference.abs_position
+        # To enable loops to work correctly we can set the start/stop point
+        # and not just use base it on the reference's absolute position
+        if start_point and not isinstance(start_point, int):
+            raise TypeError(f"The start_point passed into a "
+                            f"DefinitionUseChain must be an int but found "
+                            f"'{type(start_point).__name__}'.")
+        if stop_point and not isinstance(stop_point, int):
+            raise TypeError(f"The stop_point passed into a "
+                            f"DefinitionUseChain must be an int but found "
+                            f"'{type(stop_point).__name__}'.")
+        self._start_point = start_point
+        self._stop_point = stop_point
+        if control_flow_region is None:
+            self._scope = [reference.ancestor(Routine)]
+            if self._scope[0] is None:
+                self._scope = reference.root.children[:]
+        else:
+            # We need a list of regions for control flow.
+            if not isinstance(control_flow_region, list):
+                raise TypeError(
+                    f"The control_flow_region passed into a "
+                    f"DefinitionUseChain must be a list but found "
+                    f"'{type(control_flow_region).__name__}'."
+                )
+            self._scope = control_flow_region
+        if is_local is None:
+            # TODO Work out if symbol is local to the scope or
+            # globally accessible. Anything that isn't a local
+            # variable in a Routine has to be assumed as globally
+            # accessible for now.
+            self._is_local = isinstance(
+                reference.symbol.interface, AutomaticInterface
+            )
+        else:
+            self._is_local = is_local
+
+        # The uses, defsout and killed sets as defined for each basic block.
+        self._uses = []
+        self._defsout = []
+        self._killed = []
+
+        # The output map, mapping between nodes and the reach of that node.
+        self._reaches = []
+        self._writes = []
+
+    @property
+    def uses(self):
+        """
+        :returns: the list of nodes used by this DefinitionUseChain.
```

Comment:
"the list of nodes using the value that the referenced symbol has, before it is reassigned." if this is what it is?

Reply:
Yeah I think thats accurate.